code/modules/mining/boulder_processing.dm

```diff
@@ -319,10 +319,41 @@
 	circuit = /obj/item/circuitboard/machine/refinery
 	usage_sound = 'sound/machines/mining/refinery.ogg'
 
+	/// Reagents that we can use to wash the boulders
+	var/list/allowed_reagents = list(/datum/reagent/water, /datum/reagent/lube)
+	/// Internal beaker for storing washing fluid
+	var/obj/item/reagent_containers/cup/beaker/large/washing_input
+
 /obj/machinery/bouldertech/refinery/Initialize(mapload)
 	. = ..()
-	///We need a component like reaction chamber, but really it should be a simple demand, and then a more selective supply.
-	///It only accepts water, lube, and maybe some third chem, and should output WHEN USED, industrial waste chem to the output. Do not use any chemicals if waste output is full.
+
+	washing_input = new()
+	create_reagents(100, TRANSPARENT)
+	AddComponent(/datum/component/plumbing/selective, anchored, custom_receiver = washing_input.reagents, allowed_reagents = src.allowed_reagents)
+
+/// Try and draw reagents and produce waste
+/obj/machinery/bouldertech/refinery/proc/process_reagents(volume = MACHINE_REAGENT_TRANSFER)
```

Comment:
You'll want to call this whenever a boulder is processed (or you tried to process